File: java/Benchmark.java
import java.util.Random;

public class Benchmark {

    static final int WARMUP  =   100_000;   // discarded — lets JIT compile hot paths
    static final int MEASURE = 1_000_000;   // actual measurement

    public static void main(String[] args) {
        System.out.println("=== LOB Java Benchmark ===\n");

        // warm up — run once and discard the time
        submitBenchmark(WARMUP, false);
        cancelBenchmark(WARMUP, false);

        // measure
        submitBenchmark(MEASURE, true);
        cancelBenchmark(MEASURE, true);
    }


    // ------------------------------------------------------------------ //
    // Benchmark 1: submit() throughput                                     //
    // Measures how fast the matching engine can process incoming orders.   //
    // ------------------------------------------------------------------ //

    static void submitBenchmark(int n, boolean print) {
        OrderBook book = new OrderBook();
        Random rng = new Random(42);
        int mid = 5000;

        // seed some depth so matches actually happen during measurement
        for (int i = 0; i < 500; i++) {
            int price = Math.max(1, mid + (int)(rng.nextGaussian() * 3));
            Side side = rng.nextBoolean() ? Side.BUY : Side.SELL;
            book.submit(new Order(side, price, 10));
        }

        long start = System.nanoTime();
        for (int i = 0; i < n; i++) {
            int price = Math.max(1, mid + (int)(rng.nextGaussian() * 3));
            Side side = rng.nextBoolean() ? Side.BUY : Side.SELL;
            book.submit(new Order(side, price, 10));
        }
        long elapsed = System.nanoTime() - start;

        if (print) printResult("submit (add + match)", n, elapsed);
    }


    // ------------------------------------------------------------------ //
    // Benchmark 2: cancelOrder() throughput                               //
    // Adds N orders then cancels them all — tests the HashMap O(1) path.  //
    // ------------------------------------------------------------------ //

    static void cancelBenchmark(int n, boolean print) {
        OrderBook book = new OrderBook();
        int mid = 5000;

        // add orders at prices unlikely to match each other
        long[] ids = new long[n];
        for (int i = 0; i < n; i++) {
            // alternate BUY far below mid and SELL far above mid — no matches
            int price  = (i % 2 == 0) ? mid - 100 : mid + 100;
            Side side  = (i % 2 == 0) ? Side.BUY : Side.SELL;
            Order order = new Order(side, price, 10);
            book.addOrder(order);
            ids[i] = order.orderId;
        }

        long start = System.nanoTime();
        for (int i = 0; i < n; i++) {
            book.cancelOrder(ids[i]);
        }
        long elapsed = System.nanoTime() - start;

        if (print) printResult("cancel (O(1) lookup)", n, elapsed);
    }


    // ------------------------------------------------------------------ //
    // Formatting                                                           //
    // ------------------------------------------------------------------ //

    static void printResult(String label, int n, long elapsedNs) {
        double nsPerOp     = (double) elapsedNs / n;
        double mOpsPerSec  = n / (elapsedNs / 1e9) / 1_000_000;
        System.out.printf("%-26s  %,8d ops   %6.1f ns/op   %5.2f M ops/sec%n",
                          label, n, nsPerOp, mOpsPerSec);
    }
}

File: python/benchmark.py
"""Python equivalent of Benchmark.java — run after the Java benchmark to compare."""
import time
from random import Random
from order_book import OrderBook
from order import Order, Side

WARMUP  =   100_000
MEASURE = 1_000_000

def submit_benchmark(n, rng, print_result=True):
    book = OrderBook()
    mid  = 5000

    for _ in range(500):
        price = max(1, round(rng.gauss(mid, 3)))
        side  = rng.choice([Side.BUY, Side.SELL])
        book.submit(Order(side, price, 10))

    start = time.perf_counter_ns()
    for _ in range(n):
        price = max(1, round(rng.gauss(mid, 3)))
        side  = rng.choice([Side.BUY, Side.SELL])
        book.submit(Order(side, price, 10))
    elapsed = time.perf_counter_ns() - start

    if print_result:
        ns_per_op    = elapsed / n
        m_ops_per_sec = n / (elapsed / 1e9) / 1_000_000
        print(f"{'submit (add + match)':<26}  {n:>8,} ops   {ns_per_op:6.1f} ns/op   {m_ops_per_sec:5.2f} M ops/sec")

def cancel_benchmark(n, print_result=True):
    book = OrderBook()
    mid  = 5000

    orders = []
    for i in range(n):
        price = mid - 100 if i % 2 == 0 else mid + 100
        side  = Side.BUY  if i % 2 == 0 else Side.SELL
        o = Order(side, price, 10)
        book.add_order(o)
        orders.append(o.order_id)

    start = time.perf_counter_ns()
    for oid in orders:
        book.cancel_order(oid)
    elapsed = time.perf_counter_ns() - start

    if print_result:
        ns_per_op    = elapsed / n
        m_ops_per_sec = n / (elapsed / 1e9) / 1_000_000
        print(f"{'cancel (O(1) lookup)':<26}  {n:>8,} ops   {ns_per_op:6.1f} ns/op   {m_ops_per_sec:5.2f} M ops/sec")

if __name__ == '__main__':
    print("=== LOB Python Benchmark ===\n")
    rng = Random(42)

    # warmup
    submit_benchmark(WARMUP, rng, print_result=False)
    cancel_benchmark(WARMUP, print_result=False)

    # measure
    submit_benchmark(MEASURE, rng, print_result=True)
    cancel_benchmark(MEASURE, print_result=True)
